remotes/docker/authorizer.go

```diff
@@ -304,12 +304,15 @@ func (ah *authHandler) doBearerAuth(ctx context.Context) (token, refreshToken st
 				// As of September 2017, GCR is known to return 404.
 				// As of February 2018, JFrog Artifactory is known to return 401.
 				// As of January 2022, ACR is known to return 400.
-				if (errStatus.StatusCode == 405 && to.Username != "") || errStatus.StatusCode == 404 || errStatus.StatusCode == 401 || errStatus.StatusCode == 400 {
-					resp, err := auth.FetchToken(ctx, ah.client, ah.header, to)
-					if err != nil {
-						return "", "", err
+				switch status := errStatus.StatusCode; status {
+				case http.StatusMethodNotAllowed, http.StatusNotFound, http.StatusUnauthorized, http.StatusBadRequest:
+					if to.Username != "" {
```

Comment:
I think you change the behavior. Username != nil is only for 405. The anonymous pull comes with empty Username